### PR TITLE
Move Demos to navbar

### LINKED ISSUE #820: Move Demos to the Navbar
**Description**
The goal is to move Demos from the Knowledge Base as the nature of Demos is quite different from the Knowledge Base. The goal is then to find a better name for Knowledge Base, like Tutorials.

**Expected change**
- We should change the place of demos.
- Rename Demos to Gallery
- Add to the Navbar
- Keep the mechanism to build the demos/gallery index page
- Have all links still working

### PR DESCRIPTION
**Pull request by FlorianJacta**
Resolves: #820 

- Change the place of demos in directory,
- Rename Demos to Gallery,
- Add to the Navbar,
- Keep the mechanism to build the demos/gallery index page,
- Have all links working.

image

image

image

image

image

There are now two tiles here. We could let a Gallery tile. However, the Knowledge Base will surely change and be refactored as well. Maybe you have some comments on it.

image

### CODE REVIEW

File: docs/gallery/index.md_template [hunk 1]
```diff
@@ -3,7 +3,7 @@ hide:
   - toc
 ---
 
-Let's explore demos of applications made with Taipy.
+Let's explore applications made with Taipy.
```
Comment (jrobinAV) [inline docs/gallery/index.md_template:6]: This line is useless to me. I would remove it to gain some space. 
It's obvious to me that we are in a Gallery to explore Taipu applications.

File: tools/_setup_generation/step_gallery.py [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+    def __init__(self):
+        self.GALLERY_FOLDER_PATHS = None
+        self.GALLERY_INDEX_PATH = None
+        self.GALLERY_INDEX_TPL_PATH = None
+        self.gallery = ""
+
+
+    def enter(self, setup: Setup):
+        self.GALLERY_FOLDER_PATH = os.path.join(setup.docs_dir, self.GALLERY_FOLDER_NAME)
+        self.GALLERY_INDEX_PATH = os.path.join(setup.docs_dir, self.GALLERY_FOLDER_NAME, "index.md")
```
Comment (FabienLelaquais) [inline tools/_setup_generation/step_gallery.py]:
```suggestion
        self.GALLERY_INDEX_PATH = os.path.join(self.GALLERY_FOLDER_PATH, "index.md")
```

File: tools/_setup_generation/step_gallery.py [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+        self.GALLERY_INDEX_TEMPLATE_PATH = os.path.join(setup.docs_dir, self.GALLERY_FOLDER_NAME, "index.md_template")
```
Comment (FabienLelaquais) [inline tools/_setup_generation/step_gallery.py]:
```suggestion
        self.GALLERY_INDEX_TEMPLATE_PATH = os.path.join(self.GALLERY_FOLDER_PATH, "index.md_template")
```

File: tools/_setup_generation/step_gallery.py [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+    def setup(self, setup: Setup) -> None:
+        try:
+            # Parse the gallery folders and get all the items
+            gallery_items = self._get_list_of_items(self.GALLERY_FOLDER_PATH)
+
+            # Generate content for the kb index page
+            self.gallery = self._build_gallery_content(gallery_items)
+            self._update_gallery_index_file()
+
+            print(f"{len(gallery_items)} gallery processed.")
```
Comment (FabienLelaquais) [inline tools/_setup_generation/step_gallery.py]:
```suggestion
            print(f"{len(gallery_items)} gallery items processed.")
```

File: tools/_setup_generation/step_gallery.py (hunk 3 above)
Comment (FabienLelaquais) [reply tools/_setup_generation/step_gallery.py]: Creating a variable for the template source should not be a good idea.
You should, in the code, depend on the actual output file name (self.GALLERY_INDEX_PATH here) and add '_template' when you need it. So both file names are tightly coupled with no risk of diverging.

File: tools/_setup_generation/step_gallery.py [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+                        try:
+                            item = FolderItem(folder_path, sub_folder)
+                            items.append(item)
+                        except NoIndexFile as e:
+                            print(f"WARNING - ", e)
+                else:
+                    if not sub_folder.startswith("index.md"):
+                        item = FileItem(folder_path, sub_folder)
+                        items.append(item)
+            except NoHeader as e:
```
Comment (FabienLelaquais) [inline tools/_setup_generation/step_gallery.py]:
```suggestion
            except (NoHeader, WrongHeader) as e:
```
don't repeat the code

File: tools/_setup_generation/step_viselements.py [hunk 6]
```diff
@@ -450,25 +450,25 @@ def setup(self, setup: Setup) -> None:
         self.generate_pages("blocks", self.blocks_path)
         self.generate_builder_api()
 
-    # Special case for charts: we want to insert the chart gallery that is stored in the
+    # Special case for charts: we want to insert the chart demos that is stored in the
```
Comment (FabienLelaquais) [inline tools/_setup_generation/step_viselements.py]:
```suggestion
    # Special case for charts: we want to insert the chart gallery that is stored in the
```

File: tools/_setup_generation/step_viselements.py [hunk 7]
```diff
@@ -450,25 +450,25 @@ def setup(self, setup: Setup) -> None:
...
     # file whose path is in self.charts_home_html_path
     # This should be inserted before the first level 1 header
     def chart_page_hook(
         self, element_documentation: str, before: str, after: str, charts_md_dir: str
     ) -> tuple[str, str]:
         with open(self.charts_home_html_path, "r") as html_fragment_file:
-            chart_gallery = html_fragment_file.read()
-            # The chart_gallery begins with a comment where all sub-sections
+            chart_demos = html_fragment_file.read()
```
Comment (FabienLelaquais) [inline tools/_setup_generation/step_viselements.py]:
```suggestion
            chart_gallery = html_fragment_file.read()
```
and all the changes below

File: tools/_setup_generation/step_viselements.py
Comment (FabienLelaquais) [inline tools/_setup_generation/step_viselements.py:1]: This should NOT have been modified.

**Review by FabienLelaquais — CHANGES_REQUESTED** [review]
Modest changes, please.

File: mkdocs.yml_template
Comment (FabienLelaquais) [inline mkdocs.yml_template:1]: Did you check this addition on a mobile device screen?

File: tools/_setup_generation/step_viselements.py
Comment (jrobinAV) [inline tools/_setup_generation/step_viselements.py:1]: Do you want to change this file?
I guess it is an abusive replace all, right?

Comment (FlorianJacta) [reply tools/_setup_generation/step_viselements.py:1]: Yes, I know. I saw that I did and thought I fixed it. Thanks!

File: docs/gallery/index.md_template (hunk 1 above)
Comment (FlorianJacta) [reply docs/gallery/index.md_template:6]: Yes, I didn't want to change this PR's content. But I can change it directly here. I am also waiting on @AAstridVH to give me insights for a better UX

File: tools/_setup_generation/step_viselements.py
Comment (FlorianJacta) [reply tools/_setup_generation/step_viselements.py:1]: This is strange, I don't have them on my local branch. Everything is "demos" and not "gallery". But on the remote branch on Github, it is "gallery"...

Comment (FlorianJacta) [reply tools/_setup_generation/step_viselements.py:1]: Yes and I found it and fixed it before. I have the correct version on my local branch but not on the remote branch on Github...

**Review by jrobinAV — APPROVED** [review]

File: tools/_setup_generation/step_gallery.py (hunk 3 above)
Comment (FlorianJacta) [reply tools/_setup_generation/step_gallery.py]: I think I have corrected it now

**Review by FabienLelaquais — COMMENTED** [review]
👍

File: mkdocs.yml_template
Comment (FlorianJacta) [reply mkdocs.yml_template:1]: Tested

**Review by FabienLelaquais — APPROVED** [review]
👌

**Pull request merged**